openapi/nexus.json

```diff
@@ -6229,6 +6323,50 @@
           "items"
         ]
       },
+      "OrganizationRoles": {
+        "type": "string",
+        "enum": [
+          "Admin",
+          "Collaborator"
+        ]
+      },
+      "OrganizationRolesPolicy": {
+        "description": "Client view of a [`Policy`], which describes how this resource may be accessed\n\nNote that the Policy only describes access granted explicitly for this resource.  The policies of parent resources can also cause a user to have access to this resource.",
+        "type": "object",
+        "properties": {
+          "role_assignments": {
+            "description": "Roles directly assigned on this resource",
+            "type": "array",
+            "items": {
+              "$ref": "#/components/schemas/OrganizationRolesRoleAssignment"
+            }
+          }
+        },
+        "required": [
+          "role_assignments"
+        ]
+      },
+      "OrganizationRolesRoleAssignment": {
```

Comment:
I don't like the thought that every resource will have its own copy of these three types, though if the roles can be different I guess we have no choice?

Reply:
Yeah, I think this is a tradeoff for having the allowed roles be expressed in the API.  If there's another way to do it without that tradeoff that'd be great!

Reply:
Will all such types be the same except for their names?

Reply:
If they will, then one approach would be to `impl From<RolesPolicy> for OrganizationRolesPolicy` on all the individual types. That would give them all the same type in the OpenAPI spec.

Reply:
But I see what we're buying with the separate types: a generated client will already know at the type level what roles are valid for a given resource, which is pretty nifty.

```ts
type OrganizationRoles = 'Admin' | 'Collaborator'

type OrganizationRolesPolicy = {
  roleAssignments: OrganizationRolesRoleAssignment[]
}

type OrganizationRolesRoleAssignment = {
  identityId: string
  identityType: IdentityType
  roleName: OrganizationRoles
}
```

I guess I'm basically lamenting that OpenAPI doesn't let us express this relationship (as far as I know):

```ts
type RolesPolicy<Roles> = {
  roleAssignments: RoleAssignment<Roles>[]
}

type RoleAssignment<Roles> = {
  identityId: string
  identityType: IdentityType
  roleName: Roles
}

type OrganizationRoles = 'Admin' | 'Collaborator'
// use RoleAssignment<OrganizationRoles> wherever

type ProjectRoles = 'Admin' | 'Collaborator'
// use RoleAssignment<ProjectRoles> wherever
```

Reply:
Yeah, it would be neat if OpenAPI let you do that.

To summarize: my primary intent here was to have the Nexus type reflect the set of allowed role names.  That way, this validation is handled like all other validation, while parsing the input in the first place, and the other code paths can know from the type that the role names have been validated.

I went the logical next step of using this type directly in the HTTP handler functions (both input and output) so that the OpenAPI spec reflects the allowed values, especially since we _do_ validate that so we may as well document what the accepted values are in the spec.  As you noted, this means the generated clients can know exactly which values are allowed, which is cool.  We don't _have_ to do this part though.  We could still accept and return strings instead of enums.  And it's possible that in the future we'll want to do that because if we support custom roles, then there won't be an enum of allowed values.  But I'd rather not spend much time now worrying about that for the usual reasons: it's hard to know we're actually solving the future problem no matter what we do now, and we can always evolve this API compatibly anyway.

Reply:
@david-crespo is your lament, in essence, that OpenAPI / JSON Schema doesn't permit the expression of generics types / types with parameters?

Reply:
Yes — having identical `*RolePolicy` and `*RoleAssignment` types for every resource feels silly. But it seems there's no way around it.